pyiceberg/io/pyarrow.py

```diff
@@ -1268,14 +1265,8 @@ def __init__(self, file_schema: Schema):
 
     def _cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
         file_field = self.file_schema.find_field(field.field_id)
-        if field.field_type.is_primitive:
-            if field.field_type != file_field.field_type:
-                return values.cast(schema_to_pyarrow(promote(file_field.field_type, field.field_type), include_field_ids=False))
-            elif (target_type := schema_to_pyarrow(field.field_type, include_field_ids=False)) != values.type:
```

Comment:
The removal of this casting logic results in errors when writing the parquet files:
```
>               for data_file in _dataframe_to_data_files(table_metadata=txn.table_metadata, df=pa_table, io=txn._table.io):

tests/integration/test_writes/test_writes.py:732: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyiceberg/table/__init__.py:2944: in _dataframe_to_data_files
    yield from write_file(
/usr/local/python/3.10.13/lib/python3.10/concurrent/futures/_base.py:621: in result_iterator
    yield _result_or_cancel(fs.pop())
/usr/local/python/3.10.13/lib/python3.10/concurrent/futures/_base.py:319: in _result_or_cancel
    return fut.result(timeout)
/usr/local/python/3.10.13/lib/python3.10/concurrent/futures/_base.py:458: in result
    return self.__get_result()
/usr/local/python/3.10.13/lib/python3.10/concurrent/futures/_base.py:403: in __get_result
    raise self._exception
/usr/local/python/3.10.13/lib/python3.10/concurrent/futures/thread.py:58: in run
    result = self.fn(*self.args, **self.kwargs)
pyiceberg/io/pyarrow.py:1915: in write_parquet
    writer.write(arrow_table, row_group_size=row_group_size)
/home/codespace/.cache/pypoetry/virtualenvs/pyiceberg-FsHa-ZgB-py3.10/lib/python3.10/site-packages/pyarrow/parquet/core.py:1052: in write
    self.write_table(table_or_batch, row_group_size)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyarrow.parquet.core.ParquetWriter object at 0x7f79a96cc820>, table = pyarrow.Table
foo: string
----
foo: [["a",null,"z"]], row_group_size = 134217728

    def write_table(self, table, row_group_size=None):
        """
        Write Table to the Parquet file.
    
        Parameters
        ----------
        table : Table
        row_group_size : int, default None
            Maximum number of rows in each written row group. If None,
            the row group size will be the minimum of the Table size
            and 1024 * 1024.  If set larger than 64Mi then 64Mi will
            be used instead.
    
        """
        if self.schema_changed:
            table = _sanitize_table(table, self.schema, self.flavor)
        assert self.is_open
    
        if not table.schema.equals(self.schema, check_metadata=False):
            msg = ('Table schema does not match schema used to create file: '
                   '\ntable:\n{!s} vs. \nfile:\n{!s}'
                   .format(table.schema, self.schema))
>           raise ValueError(msg)
E           ValueError: Table schema does not match schema used to create file: 
E           table:
E           foo: string vs. 
E           file:
E           foo: large_string
E             -- field metadata --
E             PARQUET:field_id: '1'

/home/codespace/.cache/pypoetry/virtualenvs/pyiceberg-FsHa-ZgB-py3.10/lib/python3.10/site-packages/pyarrow/parquet/core.py:1094: ValueError
=============================================================================== short test summary info ================================================================================
FAILED tests/integration/test_writes/test_writes.py::test_create_table_transaction[session_catalog_hive-1] - ValueError: Table schema does not match schema used to create file: 
```

Reply:
What would worry me a bit, because in case we do upcasting (you supply a int, but the table type long), it won't include the field-id at line 1273

Reply:
That makes sense @Fokko - thank you for the catch! I think this is a [bug I introduced in this PR](https://github.com/apache/iceberg-python/commit/e61ef5770b4d73e683e2c78bebdd6c2165102a6b) when I attempted to remove the field IDs from the read schema without realizing how `to_requested_schema` was being used both for write, as well as for read. I think this could be fixed by exposing `include_field_ids` as an argument to the Visitors and `to_requested_schema`

Reply:
I just pushed a potential solution: https://github.com/apache/iceberg-python/pull/902/commits/4ca513b732b7ecc79d2369d18d0d2028ae0b8e84

This makes the abovementioned test pass on my end. We just make ensure that we write a schema that's compatible with the table schema.

Reply:
Looks good @Fokko 👍 👍 

Could we add a test where we are writing a file into an Iceberg table with sanitized schema name just to be safe?

https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1951

Reply:
I took another look at the code and I don't think this test is not strictly necessary / is an overkill